File: LeetCode/jvm/src/main/kotlin/com/sexton/leetcode/TwoSum.kt
package com.sexton.leetcode

object TwoSum {
    fun twoSum(numbers: IntArray, target: Int): IntArray {
        require(numbers.size >= 2) { "Given inputs do not have a solution" }

        val complementMap = mutableMapOf<Int, Int>()
        for (index in numbers.indices) {
            val number = numbers[index]
            val complement = target - number
            if (complementMap.containsKey(complement)) {
                return intArrayOf(complementMap[complement] ?: throw IllegalStateException("Wont happen"), index)
            }

            complementMap[number] = index
        }

        throw IllegalArgumentException("Given inputs do not have a solution")
    }
}

File: LeetCode/python/solutions/twosum.py
from typing import Dict, List


def two_sum(numbers: List[int], target: int) -> List[int]:
    compliment_map: Dict[int, int] = dict()
    for index in range(0, len(numbers)):
        number = numbers[index]
        compliment = target - number
        if compliment in compliment_map:
            return [compliment_map[compliment], index]

        compliment_map[number] = index

    raise ValueError('Given numbers does not contain the specified target')
